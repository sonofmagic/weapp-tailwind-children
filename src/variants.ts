import { getOptions } from './defaults'
import type { AddVariantParams, UserDefinedOptions } from './types'

export function createVariants (options: UserDefinedOptions = {}) {
  const { aliasList, elements, fallbackElementTags } = getOptions(options)
  const variants = elements.map((element) => {
    return aliasList.map<AddVariantParams[]>(([selector, aliases]) => {
      return aliases.map<AddVariantParams>((alias) => {
        const variant = alias + (element ? `-${element}` : '')
        const multiple = []
        if (element) {
          const base = ['&']
          selector !== ' ' && base.push(selector)
          base.push(`${element}:not(.not-${variant})`)
          multiple.push(base.join(' '))
        } else {
          for (let i = 0; i < fallbackElementTags.length; i++) {
            const e = fallbackElementTags[i]
            const base = ['&']
            selector !== ' ' && base.push(selector)
            base.push(`${e}:not(.not-${variant})`)
            multiple.push(base.join(' '))
          }
        }

        return [variant, multiple]
      })
    })
  })
  return variants
}

export const variants = createVariants()

export function each (
  variants: AddVariantParams[][][],
  cb: (params: AddVariantParams) => void
) {
  variants.forEach((v) => {
    v.forEach((x) => {
      x.forEach((y) => {
        cb(y)
      })
    })
  })
}
