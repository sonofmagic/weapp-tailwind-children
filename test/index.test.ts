import { variants, each, createVariants } from '@/variants'

describe('[Default]', () => {
  test('variants', () => {
    expect(variants).toMatchSnapshot()
  })

  test('callback', () => {
    const arr: any[] = []
    each(variants, (x) => {
      arr.push(x)
    })
    expect(arr).toMatchSnapshot()
  })

  test('createVariants', () => {
    expect(
      createVariants({
        fallbackElements: ['view']
      })
    ).toMatchSnapshot()
  })
})
