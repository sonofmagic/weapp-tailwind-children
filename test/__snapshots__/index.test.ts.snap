// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Default] callback 1`] = `
Array [
  Array [
    "children",
    Array [
      "& > view:not(.not-children)",
      "& > text:not(.not-children)",
    ],
  ],
  Array [
    "child",
    Array [
      "& > view:not(.not-child)",
      "& > text:not(.not-child)",
    ],
  ],
  Array [
    "heir",
    Array [
      "& view:not(.not-heir)",
      "& text:not(.not-heir)",
    ],
  ],
  Array [
    "descendant",
    Array [
      "& view:not(.not-descendant)",
      "& text:not(.not-descendant)",
    ],
  ],
  Array [
    "sibling",
    Array [
      "& ~ view:not(.not-sibling)",
      "& ~ text:not(.not-sibling)",
    ],
  ],
  Array [
    "twin",
    Array [
      "& ~ view:not(.not-twin)",
      "& ~ text:not(.not-twin)",
    ],
  ],
  Array [
    "next",
    Array [
      "& + view:not(.not-next)",
      "& + text:not(.not-next)",
    ],
  ],
  Array [
    "children-view",
    Array [
      "& > view:not(.not-children-view)",
    ],
  ],
  Array [
    "child-view",
    Array [
      "& > view:not(.not-child-view)",
    ],
  ],
  Array [
    "heir-view",
    Array [
      "& view:not(.not-heir-view)",
    ],
  ],
  Array [
    "descendant-view",
    Array [
      "& view:not(.not-descendant-view)",
    ],
  ],
  Array [
    "sibling-view",
    Array [
      "& ~ view:not(.not-sibling-view)",
    ],
  ],
  Array [
    "twin-view",
    Array [
      "& ~ view:not(.not-twin-view)",
    ],
  ],
  Array [
    "next-view",
    Array [
      "& + view:not(.not-next-view)",
    ],
  ],
  Array [
    "children-text",
    Array [
      "& > text:not(.not-children-text)",
    ],
  ],
  Array [
    "child-text",
    Array [
      "& > text:not(.not-child-text)",
    ],
  ],
  Array [
    "heir-text",
    Array [
      "& text:not(.not-heir-text)",
    ],
  ],
  Array [
    "descendant-text",
    Array [
      "& text:not(.not-descendant-text)",
    ],
  ],
  Array [
    "sibling-text",
    Array [
      "& ~ text:not(.not-sibling-text)",
    ],
  ],
  Array [
    "twin-text",
    Array [
      "& ~ text:not(.not-twin-text)",
    ],
  ],
  Array [
    "next-text",
    Array [
      "& + text:not(.not-next-text)",
    ],
  ],
]
`;

exports[`[Default] createVariants 1`] = `
Array [
  Array [
    Array [
      Array [
        "children",
        Array [
          "& > view:not(.not-children)",
        ],
      ],
      Array [
        "child",
        Array [
          "& > view:not(.not-child)",
        ],
      ],
    ],
    Array [
      Array [
        "heir",
        Array [
          "& view:not(.not-heir)",
        ],
      ],
      Array [
        "descendant",
        Array [
          "& view:not(.not-descendant)",
        ],
      ],
    ],
    Array [
      Array [
        "sibling",
        Array [
          "& ~ view:not(.not-sibling)",
        ],
      ],
      Array [
        "twin",
        Array [
          "& ~ view:not(.not-twin)",
        ],
      ],
    ],
    Array [
      Array [
        "next",
        Array [
          "& + view:not(.not-next)",
        ],
      ],
    ],
  ],
  Array [
    Array [
      Array [
        "children-view",
        Array [
          "& > view:not(.not-children-view)",
        ],
      ],
      Array [
        "child-view",
        Array [
          "& > view:not(.not-child-view)",
        ],
      ],
    ],
    Array [
      Array [
        "heir-view",
        Array [
          "& view:not(.not-heir-view)",
        ],
      ],
      Array [
        "descendant-view",
        Array [
          "& view:not(.not-descendant-view)",
        ],
      ],
    ],
    Array [
      Array [
        "sibling-view",
        Array [
          "& ~ view:not(.not-sibling-view)",
        ],
      ],
      Array [
        "twin-view",
        Array [
          "& ~ view:not(.not-twin-view)",
        ],
      ],
    ],
    Array [
      Array [
        "next-view",
        Array [
          "& + view:not(.not-next-view)",
        ],
      ],
    ],
  ],
]
`;

exports[`[Default] variants 1`] = `
Array [
  Array [
    Array [
      Array [
        "children",
        Array [
          "& > view:not(.not-children)",
          "& > text:not(.not-children)",
        ],
      ],
      Array [
        "child",
        Array [
          "& > view:not(.not-child)",
          "& > text:not(.not-child)",
        ],
      ],
    ],
    Array [
      Array [
        "heir",
        Array [
          "& view:not(.not-heir)",
          "& text:not(.not-heir)",
        ],
      ],
      Array [
        "descendant",
        Array [
          "& view:not(.not-descendant)",
          "& text:not(.not-descendant)",
        ],
      ],
    ],
    Array [
      Array [
        "sibling",
        Array [
          "& ~ view:not(.not-sibling)",
          "& ~ text:not(.not-sibling)",
        ],
      ],
      Array [
        "twin",
        Array [
          "& ~ view:not(.not-twin)",
          "& ~ text:not(.not-twin)",
        ],
      ],
    ],
    Array [
      Array [
        "next",
        Array [
          "& + view:not(.not-next)",
          "& + text:not(.not-next)",
        ],
      ],
    ],
  ],
  Array [
    Array [
      Array [
        "children-view",
        Array [
          "& > view:not(.not-children-view)",
        ],
      ],
      Array [
        "child-view",
        Array [
          "& > view:not(.not-child-view)",
        ],
      ],
    ],
    Array [
      Array [
        "heir-view",
        Array [
          "& view:not(.not-heir-view)",
        ],
      ],
      Array [
        "descendant-view",
        Array [
          "& view:not(.not-descendant-view)",
        ],
      ],
    ],
    Array [
      Array [
        "sibling-view",
        Array [
          "& ~ view:not(.not-sibling-view)",
        ],
      ],
      Array [
        "twin-view",
        Array [
          "& ~ view:not(.not-twin-view)",
        ],
      ],
    ],
    Array [
      Array [
        "next-view",
        Array [
          "& + view:not(.not-next-view)",
        ],
      ],
    ],
  ],
  Array [
    Array [
      Array [
        "children-text",
        Array [
          "& > text:not(.not-children-text)",
        ],
      ],
      Array [
        "child-text",
        Array [
          "& > text:not(.not-child-text)",
        ],
      ],
    ],
    Array [
      Array [
        "heir-text",
        Array [
          "& text:not(.not-heir-text)",
        ],
      ],
      Array [
        "descendant-text",
        Array [
          "& text:not(.not-descendant-text)",
        ],
      ],
    ],
    Array [
      Array [
        "sibling-text",
        Array [
          "& ~ text:not(.not-sibling-text)",
        ],
      ],
      Array [
        "twin-text",
        Array [
          "& ~ text:not(.not-twin-text)",
        ],
      ],
    ],
    Array [
      Array [
        "next-text",
        Array [
          "& + text:not(.not-next-text)",
        ],
      ],
    ],
  ],
]
`;
